src/components/schedule-visualizer/index.tsx

```diff
@@ -215,23 +220,45 @@ export default function ScheduleVisualizer(props: ScheduleVisualizerProps): Reac
     setCurLevelName(e.name);
   }
 
-  function getTrajectory(levelName: string): Trajectory[] {
-    const resp = trajectories[levelName];
-    return resp ? resp.values : [];
-  }
-
-  function getConflicts(levelName: string): Conflict[] {
-    const resp = trajectories[levelName];
-    return resp ? resp.conflicts : [];
-  }
-
   const sortedMapFloorLayers = mapFloorLayerSort.map(x => mapFloorLayers[x]);
   const ref = React.useRef<ImageOverlay>(null);
 
   if (ref.current) {
     ref.current.leafletElement.setZIndex(0);
   }
 
+  React.useEffect(() => {
+    function getTrajectory(levelName: string): Trajectory[] {
+      const resp = trajectories[levelName];
+      return resp ? resp.values : [];
+    }
+
+    function getConflicts(levelName: string): Conflict[] {
+      const resp = trajectories[levelName];
+      return resp ? resp.conflicts : [];
+    }
+
+    function getConflictRobotsName(conflicts: Conflict[], trajs: Trajectory[]): string[][] {
+      let conflictRobotNames: string[][] = [];
+      if (conflicts.length !== 0) {
+        conflicts.forEach(conflictPair => {
+          let robotNames: string[] = [];
+          conflictPair.forEach(conflictId => {
+            const robotName = DefaultTrajectoryManager.getRobotNameFromPathId(conflictId, trajs);
+            robotName && robotNames.push(robotName);
+          });
+          robotNames && conflictRobotNames.push(robotNames);
+        });
+      }
+      return conflictRobotNames;
+    }
+    if (curMapFloorLayer) {
+      setCurMapTrajectories(getTrajectory(curMapFloorLayer.level.name));
+      setCurMapConflicts(getConflicts(curMapFloorLayer.level.name));
+      setConflictRobotNames(getConflictRobotsName(curMapConflicts, curMapTrajectories));
```

Comment:
I would be extremely careful with this because setState is an async function so maybe `getConflictRobotsName(curMapConflicts, curMapTrajectories) ` can be executed with old values.

Reply:
Can you do something like this?

``` typescript
 const mapTrajectories = getTrajectory(curMapFloorLayer.level.name);
 const mapConflicts = getConflicts(curMapFloorLayer.level.name);
 setCurMapTrajectories(mapTrajectories);
 setCurMapConflicts(mapConflicts);
 setConflictRobotNames(getConflictRobotsName(mapConflicts, mapTrajectories));
```

To make sure we are passing the updated values to `setConflictRobotNames`?